### PR TITLE
[WIP] convert branch protector to trusted job

### PR DESCRIPTION
**Pull request by BenTheElder**
fixes https://github.com/kubernetes/test-infra/issues/10015

TODO:
-  [ ] the job-config configmap (?)
- [ ] remove the old k8s cron
- [ ] update the docs / add an announcement for other prow users

### CODE REVIEW

**Comment by BenTheElder** [comment]
> I assume you know how to translate this into a docker run --rm=true incantation?

I mean yes, but why would we need to do that?

Also uhhh https://gubernator.k8s.io/builds/kubernetes-jenkins/pr-logs/pull/test-infra/10016/pull-test-infra-verify-codegen/ passed just fine but prow says it failed (!?) @cjwagner any idea how that might have happened?

**Comment by BenTheElder** [comment]
finished.json is SUCCESS and the build log is there.... wat?

**Comment by cjwagner** [comment]
> Also uhhh https://gubernator.k8s.io/builds/kubernetes-jenkins/pr-logs/pull/test-infra/10016/pull-test-infra-verify-codegen/ passed just fine but prow says it failed (!?) @cjwagner any idea how that might have happened?

The pod must have completed with a failed status (a container exited non-zero) after the tests passed and `finished.json` was successfully uploaded.

**Comment by fejta** [comment]
> why would we need to do that

To run the job locally before creating the PR

**Comment by BenTheElder** [comment]
@cjwagner Unfortunately I'm having some difficulty getting the pod logs... this is pretty surprising though. Why would the prowjob be failed after we uploaded all artifacts (AFAICT) and the job container exited success? (and we wrote the metadata files too!)

**Comment by cjwagner** [comment]
The pod has been GCed already. We have configured a 30 minute max pod age for completed prowjobs.

File: config/jobs/kubernetes/test-infra/test-infra-trusted.yaml [hunk 1]
```diff
@@ -68,3 +68,38 @@ periodics:
...
         secretName: oauth-token
+- cron: "54 * * * *"  # Every hour at 54 minutes past the hour 
+  name: ci-branchprotector
+  cluster: test-infra-trusted
+  decorate: true
+  spec: 
+    containers:
+    - name: branchprotector
+      image: gcr.io/k8s-prow/branchprotector:v20181019-08e9d55c9
+      args:
```
Comment (BenTheElder) [inline config/jobs/kubernetes/test-infra/test-infra-trusted.yaml]: TODO: this also needs to set the `command:` to match the entrypoint binary command.

Imo, this is an annoying thing to need to do manually, really our entrypoint binary hack should respect the original image `ENTRYPOINT` if `command` isn't set by the user, like k8s

**Comment by BenTheElder** [comment]
> To run the job locally before creating the PR, aka any docker run command you can hand me that passes your test I can immediately translate into a passing decorated prow job

Isn't this https://github.com/kubernetes/test-infra/issues/6590 ? Seems a bit wider in scope than this PR... In this particular PR we'd also probably want to special case handling the job configmap.

> The pod has been GCed already. We have configured a 30 minute max pod age for completed prowjobs.

Yes, which only matters for the kubectl logs though, there should still be info in stackdriver AFAIK, but getting the log link back to a pj id is problematic.

File: config/jobs/kubernetes/test-infra/test-infra-trusted.yaml (hunk 1 above)
Comment (fejta) [reply config/jobs/kubernetes/test-infra/test-infra-trusted.yaml]: 100% agree but it is a current design limitation:

* injecting the entrypoint binary as the ENTRYPOINT is how we: a) capture/upload logs from the test process and b) notify the sidecar that the test process has finished
* determining the ENTRYPOINT of a random image is non-trivial, too expensive for trigger/decorate to download every image to extract the tag whenever it schedules a job

**Comment by cjwagner** [comment]
> Why would the prowjob be failed after we uploaded all artifacts (AFAICT) and the job container exited success? (and we wrote the metadata files too!)

```
{
  component:  "sidecar"   
  error:  "failed to upload to GCS: failed to upload to GCS: encountered errors during upload: [[Post https://www.googleapis.com/upload/storage/v1/b/kubernetes-jenkins/o?alt=json&projection=full&uploadType=multipart: oauth2: cannot fetch token: Post https://oauth2.googleapis.com/token: net/http: TLS handshake timeout] [Post https://www.googleapis.com/upload/storage/v1/b/kubernetes-jenkins/o?alt=json&projection=full&uploadType=multipart: oauth2: cannot fetch token: Post https://oauth2.googleapis.com/token: net/http: TLS handshake timeout]]"   
  level:  "fatal"   
  msg:  "Failed to report job status"   
 }
```

**Comment by BenTheElder** [comment]
And yet it wrote a `finished.json`? I can't tell from that error what it failed to write, but GCS seems to have everything you would expect http://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/test-infra/10016/pull-test-infra-verify-codegen/2994/

**Comment by cjwagner** [comment]
Thats the only error I see. Other files uploaded successfully. e.g.
```
{
  component:  "sidecar"   
  dest:  "pr-logs/pull/test-infra/10016/pull-test-infra-verify-codegen/2994/finished.json"   
  level:  "info"   
  msg:  "Finished upload"   
 }
```

**Comment by fejta** [comment]
> Isn't this #6590 ? Seems a bit wider in scope than this PR... In this particular PR we'd also probably want to special case handling the job configmap.

Unrelated

This thread isn't anything this PR needs to do, just a practice I have of running the new jobs I write locally, and noting that this is REALLY easy to run decorated prowjobs locally since there's a trivial translation between the job's spec and a docker run command: `docker run --rm=true -v $PWD:/foo -w /foo $OTHER_VOLUME_MOUNTS $ENV_SETTINGS IMAGE COMMAND ARGS`

See https://github.com/kubernetes/test-infra/pull/9895#discussion_r229428201

At some point I intend to write a tool that does this translation

File: config/jobs/kubernetes/test-infra/test-infra-trusted.yaml [hunk 2]
```diff
@@ -68,3 +68,38 @@ periodics:
...
       secret:
         secretName: oauth-token
+- cron: "54 * * * *"  # Every hour at 54 minutes past the hour 
+  name: ci-branchprotector
+  cluster: test-infra-trusted
+  decorate: true
+  spec: 
+    containers:
+    - name: branchprotector
+      image: gcr.io/k8s-prow/branchprotector:v20181019-08e9d55c9
```
Comment (fejta) [inline config/jobs/kubernetes/test-infra/test-infra-trusted.yaml]: An alternative here would be to give it a bazel image (launcher.gcr.io/google/bazel), check out test-infra and `bazel run //prow/cmd:branchprotector` then we don't need to worry about publishing images.

File: config/jobs/kubernetes/test-infra/test-infra-trusted.yaml (hunk 1 above)
Comment (BenTheElder) [reply config/jobs/kubernetes/test-infra/test-infra-trusted.yaml]: Technically we don't actually have to download the image, we just need the image manifest which is small, but I'm not sure if there's a nice library for this. Not to mention image pull secrets etc...

Agree this is a current design limitation, but I do think it makes it unnecessarily clunky to port "I have a container" to -> I now have a job.

File: config/jobs/kubernetes/test-infra/test-infra-trusted.yaml (hunk 2 above)
Comment (BenTheElder) [reply config/jobs/kubernetes/test-infra/test-infra-trusted.yaml]: SGTM
/hold

**Comment by BenTheElder** [comment]
we need to think more about how to deal with the job configmap in particular, and my bandwidth is pretty much out. we can revisit in a clean PR

/close

**Pull request closed without merging**